### PR TITLE
Fix bug in mem_eff kernel with attention mask and MQA

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

Found using the repros mentioned in this issue: #112577

After many go rounds with compute-sanitizer and eventual printf debugging I feel pretty confident that this was the underlying issue 

cc @ptrblck

### CODE REVIEW

**Comment by drisspg** [comment]
@danthe3rd Would you mind taking a look at this?

https://gist.github.com/drisspg/a9269f422d13a67ee94610f8246951a8
I verified the fix using compute-sanitizer, previously this would produce 26 errors and now I am getting:

```Shell
❯ compute-sanitizer --tool memcheck python ../scripts/sdpa/hugging_face_bug.py
========= COMPUTE-SANITIZER
WARNING:2023-12-20 17:13:24 2335518:2335518 init.cpp:155] function cbapi->getCuptiStatus() failed with error CUPTI_ERROR_MULTIPLE_SUBSCRIBERS_NOT_SUPPORTED (39)
WARNING:2023-12-20 17:13:24 2335518:2335518 init.cpp:156] CUPTI initialization failed - CUDA profiler activities will be missing
INFO:2023-12-20 17:13:24 2335518:2335518 init.cpp:158] If you see CUPTI_ERROR_INSUFFICIENT_PRIVILEGES, refer to https://developer.nvidia.com/nvidia-development-tools-solutions-err-nvgpuctrperm-cupti
/home/drisspg/meta/transformers/src/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
/home/drisspg/meta/transformers/src/transformers/utils/generic.py:309: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
Setting `pad_token_id` to `eos_token_id`:49152 for open-end generation.
---- math
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[0]
        left = []
        right = []
        for i in arr:
            if i < pivot:
                left.append(i)
            else:
                right.append(
---- math
<|endoftext|><|endoftext|><|endoftext|>a = 0; a < 10; a++) {
            for (int b = 0; b < 10; b++) {
                for (int c = 0; c < 10; c++) {
                    for (int d = 0; d < 
---- math
<|endoftext|><|endoftext|>print('I am a child of'+ parent.name)

class Parent(Person):
    def __init__(self, name):
        super().__init__(name)
        self.children = []

    def add_child(self, child):
        self.children.append(child)
        child
Setting `pad_token_id` to `eos_token_id`:49152 for open-end generation.
---- mem-efficient
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    else:
        pivot = arr[0]
        left = []
        right = []
        for i in arr:
            if i < pivot:
                left.append(i)
            else:
                right.append(
---- mem-efficient
<|endoftext|><|endoftext|><|endoftext|>a = 0; a < 10; a++) {
            for (int b = 0; b < 10; b++) {
                for (int c = 0; c < 10; c++) {
                    for (int d = 0; d < 
---- mem-efficient
<|endoftext|><|endoftext|>print('I am a child of'+ parent.name)

class Parent(Person):
    def __init__(self, name):
        super().__init__(name)
        self.children = []

    def add_child(self, child):
        self.children.append(child)
        child
========= ERROR SUMMARY: 0 errors
```

**Review by malfet — APPROVED** [review]

**Review by danthe3rd — APPROVED** [review]
Thanks a lot! Excellent debugging and bug catching there!

**Review by atalman — APPROVED** [review]
